lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -229,32 +224,54 @@ bool ProxyRenderDelegate::requiresUpdate(const MSubSceneContainer& container, co
     return true;
 }
 
+void ProxyRenderDelegate::_ClearRenderDelegate()
+{
+    // The order of deletion matters. Some orders cause crashes.
+    // This order matches the deletion order when a ProxyRenderDelegate is destroyed.
+    // Class member variables are destroyed in the reverse order in which they appear in the class declaration.
+
+    _sceneDelegate.reset(nullptr);
+    _taskController.reset(nullptr);
+    _renderIndex.reset(nullptr);
+    _renderDelegate.reset(nullptr);
+}
+
 //! \brief  One time initialization of this drawing routine
-void ProxyRenderDelegate::_InitRenderDelegate() {
-    // No need to run all the checks if we got till the end
-    if (_isInitialized())
-        return;
+void ProxyRenderDelegate::_InitRenderDelegate(MSubSceneContainer& container) {
 
     if (_proxyShape == nullptr)
         return;
 
-    if (!_usdStage) {
-        _usdStage = _proxyShape->getUsdStage();
+    UsdStageRefPtr newUsdStage = _proxyShape->getUsdStage();
+    if (_usdStage != newUsdStage)
+    {
+        // delete everything so we stop drawing the old stage and draw the new one
+        _ClearRenderDelegate();
+        _dummyTasks.clear();
+
+        _isPopulated = false;
+        container.clear();
+
+        _usdStage = newUsdStage;
     }
+    
+    // No need to run all the checks if we got till the end
+    if (_isInitialized())
+        return;
 
     if (!_renderDelegate) {
         MProfilingScope subProfilingScope(HdVP2RenderDelegate::sProfilerCategory,
             MProfiler::kColorD_L1, "Allocate VP2RenderDelegate");
-        _renderDelegate = new HdVP2RenderDelegate(*this);
+        _renderDelegate.reset(new HdVP2RenderDelegate(*this));
     }
 
     if (!_renderIndex) {
         MProfilingScope subProfilingScope(HdVP2RenderDelegate::sProfilerCategory,
             MProfiler::kColorD_L1, "Allocate RenderIndex");
 #if USD_VERSION_NUM > 2002
-        _renderIndex = HdRenderIndex::New(_renderDelegate, HdDriverVector());
+        _renderIndex.reset(HdRenderIndex::New(_renderDelegate.get(), HdDriverVector()));
 #else
-        _renderIndex = HdRenderIndex::New(_renderDelegate);
+        _renderIndex.reset(HdRenderIndex::New(_renderDelegate));
```

Comment:
This is a bit different from Line 272. Is it intentional?

Reply:
No, I didn't test against building against an USD with version num <= 2002.